### ISSUE TITLE
[REQUEST] Support for XLA/TPU

### ISSUE THREAD

**Issue by radna0** [deepspeedai/DeepSpeed#6901]
**Is your feature request related to a problem? Please describe.**
Currently, DeepSpeed lacks support for TPUs via the XLA backend. This limits the use of DeepSpeed's advanced parallelism techniques, such as pipeline parallelism and ZeRO optimizations, for TPU users. Frameworks like PyTorch/XLA and Accelerate offer TPU support, but they lack the comprehensive optimization features that DeepSpeed provides.

This is particularly frustrating for users who want to scale models efficiently on TPUs across multiple nodes while leveraging DeepSpeed's features.

**Describe the solution you'd like**
I propose integrating XLA as a backend for DeepSpeed, enabling TPU users to take advantage of DeepSpeed's optimizations, including pipeline parallelism, ZeRO, and advanced scheduling mechanisms.

**Describe alternatives you've considered**
- PyTorch/XLA + Torchrun: While it provides basic TPU support, it doesn't include DeepSpeed's advanced features.
- HuggingFace Accelerate: Offers TPU support but lacks pipeline parallelism and other optimization techniques available in DeepSpeed.

**Additional context**
There is growing interest in having TPU support in DeepSpeed, as evidenced by multiple community requests. Adding XLA as a backend would make DeepSpeed accessible to a wider audience, particularly researchers and engineers working with TPUs.

I'm willing to lead this integration if needed, and I hope this request sparks discussion and collaboration within the DeepSpeed team.

Link to Pytorch XLA Feature Request: https://github.com/pytorch/xla/issues/8514#issue-2753774566

**Comment by tjruwase** [comment]
@radna0, this sounds exciting. It is awesome that you are willing to lead the integration. We will be glad to collaborate with you to make this happen.

**Comment by radna0** [comment]
Thank you @tjruwase . I currently have a setup somewhat running. Here are some things I did and some roadblocks I have.

- Basics:
I added `XLA_Accelerator` class within `xla_accelerator.py`, this then will be used in `get_accelerator()` in `real_accelerator.py`. All the apis for the basics such as device, communication backend, metrics, RNG APIs, etc. are using Pytorch XLA and works with TPUs. But the ops are using ones from the CPU, this works fine I think and the xla ops can be added later. Right now I just want something minimal to get working with Pytorch XLA on TPUs.

- Roadblock 1:
When it comes to the DeepSpeed launcher. XLA only allows one single parent process when doing muiti process spawn, but when using DeepSpeed launcher, For example `deepspeed train.py`, it first runs the `runner.py` file and then calls `launcher.py` to spawn in the processes. XLA will see the first process here as runner,py. So here is the workaround that I have.

Run with: `DS_ACCELERATOR=cpuxla deepspeed train.py`.
What it does is, sets the `DS_ACCELERATOR` to `cpu` within `runner.py`. Before running `launcher.py` from `runner.py`, we change `DS_ACCELERATOR` to `xla`. Now this way, the parent process will be `launcher.py` instead of `runner.py`.

so it would just act/behave like the following for `DS_ACCELERATOR=cpuxla`
DS_ACCELERATOR=cpu python deepspeed/runner.py
DS_ACCELERATOR=xla python deepspeed/launcher.py

- Roadblock 2:
After getting to launcher.py, the processes are spawned with `subprocess.Popen`. This is also not compatible with Pytorch XLA, as only one process can access the devices. The only way seems to be using the `torch_xla.distributed.xla_multiprocessing.spawn`. I don't know how `subprocess.Popen` differ with torch_xla spawn method, but Popen seems to spawn an entire child process, while torch_xla spawn seems to fork the parent process. 

I haven't tested out other methods more clearly like `os.fork`. But it is stated by the pytorch xla team within their documentation that multiprocess can only be spawned with the pytorch xla spawn method.


I'm currently testing a training script to see if multiprocess, pipeline parallelism, etc works properly, but will update if I have any findings.

**Comment by tjruwase** [comment]
> Run with: `DS_ACCELERATOR=cpuxla deepspeed train.py`.
> What it does is, sets the `DS_ACCELERATOR` to `cpu` within `runner.py`. Before running `launcher.py` from `runner.py`, we change `DS_ACCELERATOR` to `xla`. Now this way, the parent process will be `launcher.py` instead of `runner.py`.

I am curious about the implementation of the above. Are you handling when `DS_ACCELERATOR` is not [available](https://github.com/microsoft/DeepSpeed/blob/eea5304807c6a04d0f2c55cb935ec295235d9b54/accelerator/real_accelerator.py#L107-L108)?

I assume you have seen these tutorials:
1. https://www.deepspeed.ai/tutorials/accelerator-abstraction-interface/
2. https://www.deepspeed.ai/tutorials/accelerator-setup-guide/

**Comment by tjruwase** [comment]
> I don't know how `subprocess.Popen` differ with torch_xla spawn method, but Popen seems to spawn an entire child process, while torch_xla spawn seems to fork the parent process.

Does it make sense to add the spawn method into the accelerator abstraction [class](https://github.com/microsoft/DeepSpeed/blob/eea5304807c6a04d0f2c55cb935ec295235d9b54/accelerator/abstract_accelerator.py#L10)?

**Comment by tjruwase** [comment]
Another idea for roadblock 1 is to add `DS_ACCELERATOR=xla` into `.deepspeed_env`, rather than setting on the command line. Please see this [description](https://www.deepspeed.ai/getting-started/#multi-node-environment-variables) of `.deepspeed_env` usage.

**Comment by radna0** [comment]
> > Run with: `DS_ACCELERATOR=cpuxla deepspeed train.py`.
> > What it does is, sets the `DS_ACCELERATOR` to `cpu` within `runner.py`. Before running `launcher.py` from `runner.py`, we change `DS_ACCELERATOR` to `xla`. Now this way, the parent process will be `launcher.py` instead of `runner.py`.
> 
> I am curious about the implementation of the above. Are you handling when `DS_ACCELERATOR` is not [available](https://github.com/microsoft/DeepSpeed/blob/eea5304807c6a04d0f2c55cb935ec295235d9b54/accelerator/real_accelerator.py#L107-L108)?
> 
> I assume you have seen these tutorials:
> 
> 1. https://www.deepspeed.ai/tutorials/accelerator-abstraction-interface/
> 2. https://www.deepspeed.ai/tutorials/accelerator-setup-guide/

I haven't seen those tutorials. But my implementation of the xla_accelerator is based on the CPU and cuda accelerator. So it is actually compatible and working just as the tutorial suggested.

As for the setup guide for Pytorch XLA, I can document it as I open my PR to merge the changes.  Or I can do that right away if needed? How can I do so?

<br>

> > I don't know how `subprocess.Popen` differ with torch_xla spawn method, but Popen seems to spawn an entire child process, while torch_xla spawn seems to fork the parent process.
> 
> Does it make sense to add the spawn method into the accelerator abstraction [class](https://github.com/microsoft/DeepSpeed/blob/eea5304807c6a04d0f2c55cb935ec295235d9b54/accelerator/abstract_accelerator.py#L10)?

Yes, I think this would be ideal, I'm currently just checking the accelerator on the fly within the launcher.py to handle spawning processes accordingly.

<br>

> Another idea for roadblock 1 is to add `DS_ACCELERATOR=xla` into `.deepspeed_env`, rather than setting on the command line. Please see this [description](https://www.deepspeed.ai/getting-started/#multi-node-environment-variables) of `.deepspeed_env` usage.


I'm not sure if I'm doing it correctly, I created `.deepspeed_env` file in the local directory, the content is only `DS_ACCELERATOR=xla`. But it doesn't seem to work launching with both these commands. It only uses CPU as the accelerator 
`deepspeed train.py`
`DS_ENV_FILE=path_to_env/.deepspeed_env deepspeed train.py`

**Comment by tjruwase** [comment]
> it doesn't seem to work launching with both these commands. It only uses CPU as the accelerator

Hmm, that is strange. Can you examine this code to see what is wrong?
https://github.com/microsoft/DeepSpeed/blob/eea5304807c6a04d0f2c55cb935ec295235d9b54/deepspeed/launcher/runner.py#L593-L600

**Comment by tjruwase** [comment]
> * `deepspeed train.py`
> * `DS_ENV_FILE=path_to_env/.deepspeed_env deepspeed train.py`

Try add `--force_multi` command line argument, i.e., `deepspeed --force_multi ...`

**Comment by radna0** [comment]
@tjruwase I tried all kinds of commands, all just uses CPU, and with `--force_multi` it results in connection error. But I don't think this ENV check should just be in multi node setting. Can we have that for both multi_node_exec and not?

Also I'm leaving out the xla check if accelerator_name is None. As it needs the CPU within the first process, runner.py. and then launcher.py can use xla. that checks then breaks if DS_ACCELERATOR is not set. which should work with CPU my default in my opinion

**Comment by tjruwase** [comment]
> [@tjruwase](https://github.com/tjruwase) I tried all kinds of commands, all just uses CPU, and with `--force_multi` it results in connection error. But I don't think this ENV check should just be in multi node setting. Can we have that for both `multi_node_exec` and not?

Agreed. Will fix.

**Comment by tjruwase** [comment]
> Also I'm leaving out the xla check `if accelerator_name is None`. As it needs the CPU within the first process, runner.py. and then launcher.py can use xla. that checks then breaks if `DS_ACCELERATOR=cpuxla` is not set. which should work with CPU by default in my opinion.

CPU should be the fall-through if no accelerator is detected or specified. So, it seems your use case is exposing an issue here. Is it possible to share a stack trace of the failure?

**Comment by tjruwase** [comment]
> so it would just act/behave like the following for `DS_ACCELERATOR=cpuxla`
> DS_ACCELERATOR=cpu python deepspeed/runner.py

On further thought, I think the above strategy will create problems in `runner.py` for logic such as 
1. [This](https://github.com/microsoft/DeepSpeed/blob/eea5304807c6a04d0f2c55cb935ec295235d9b54/deepspeed/launcher/runner.py#L428)
2. [This](https://github.com/microsoft/DeepSpeed/blob/eea5304807c6a04d0f2c55cb935ec295235d9b54/deepspeed/launcher/runner.py#L448)
3. [This](https://github.com/microsoft/DeepSpeed/blob/eea5304807c6a04d0f2c55cb935ec295235d9b54/deepspeed/launcher/runner.py#L587)

Something to keep in mind.

**Comment by radna0** [comment]
> > Also I'm leaving out the xla check `if accelerator_name is None`. As it needs the CPU within the first process, runner.py. and then launcher.py can use xla. that checks then breaks if `DS_ACCELERATOR=cpuxla` is not set. which should work with CPU by default in my opinion.
> 
> CPU should be the fall-through if no accelerator is detected or specified. So, it seems your use case is exposing an issue here. Is it possible to share a stack trace of the failure?

There are two ways can set accelerator_name:
*Note torch_xla has no checks for available device by default, we rely on import error
- setting accelerator_name = "xla" => We hit `Roadblock 1`. runner,py can only use CPU for xla accelerator to work, if not we can't access the xla device at all outside of runner.py

- setting accelerator_name = "cpuxla", will not work as well, because it only sets accelerator for get_accelerator, maybe if accelerator_name is None, we can also set the global DS_ACCELERATOR env. If this happens then 'cpuxla' will work as I stated above

`        
# TODO: This will not work right now, only with cpuxla env  
# cpu first for runner.py and xla set for launcher.py 
if accelerator_name is None:
    try:
        import torch_xla  # noqa: F401,F811

        # torch_xla has no check such as torch_xla.is_available()
        accelerator_name = "cpuxla" # or "xla"
    except ImportError as e:
        pass
`

---


> > so it would just act/behave like the following for `DS_ACCELERATOR=cpuxla`
> > DS_ACCELERATOR=cpu python deepspeed/runner.py
> 
> On further thought, I think the above strategy will create problems in `runner.py` for logic such as
> 
> 1. [This](https://github.com/microsoft/DeepSpeed/blob/eea5304807c6a04d0f2c55cb935ec295235d9b54/deepspeed/launcher/runner.py#L428)
> 2. [This](https://github.com/microsoft/DeepSpeed/blob/eea5304807c6a04d0f2c55cb935ec295235d9b54/deepspeed/launcher/runner.py#L448)
> 3. [This](https://github.com/microsoft/DeepSpeed/blob/eea5304807c6a04d0f2c55cb935ec295235d9b54/deepspeed/launcher/runner.py#L587)
> 
> Something to keep in mind.

With `DS_ACCELERATOR=cpuxla`
1. `visible_devices_envs()` does not affect how `cpuxla` works, I see that when running this way, it sets the `CUDA_VISIBLE_DEVICES` but everything runs fine.
2. `device_count()` will be a problem here, as `cpuxla` relies on the `cpu_accelerator.py`. I believe any pytorch xla imports will initializes it as the parent process. This though can resolved by having `cpuxla_accelerator.py`, which still use most of basics API by CPU, but device_count needs to be found through its metrics package [tpu-info](https://github.com/AI-Hypercomputer/cloud-accelerator-diagnostics/tree/main/tpu_info). Its has apis for detecting basics info like device and memory usage without initializing the xla runtime.
3. `export_envs()` also does not affect this. I could potentially set `PJRT_DEVICE=TPU` env, but XLA works with GPU as well. Also the env will be automatically detected by pytorch xla.

**Comment by tjruwase** [comment]
@radna0, how is this going?

**Comment by radna0** [comment]
@tjruwase I'm currently testing this simple training script here from [DeepSpeedExamples/alexnet](https://github.com/microsoft/DeepSpeedExamples/blob/master/training/pipeline_parallelism/train.py), I'm running into some problems with communicating and synchronizing, but it's mostly problems with torch_xla and how TPUs are configured by nature

**Comment by tjruwase** [comment]
Thanks for the update, @radna0. It sounds like single device training not requiring communication should be working, which is awesome! We are very invested in this effort so please let us know how we can assist. I am refreshing myself on our earlier discussions to execute action items and think more on design issues. Thanks!

**Comment by radna0** [comment]
Yes @tjruwase, training on a single device without communication is very straightforward, though I don't think that's recommended as with TPUs relating to `Roadblock 2`, when a process is spawned with TPU/XLA enabled, only that process can access the devices, no other/concurrent process can access the devices, even if the TPU device/chip is different.

Example: A TPU v3-8 system has 4 chips, each chip has 2 devices, each device has 16GB, so a total 8 devices/128GB.
```
from tpu_info import device
chip_type, count = device.get_local_chips()
chip_type: <TpuChip.V3: Info(name='v3', hbm_gib=16, devices_per_chip=2)>
count: 4
```

Problem: Process A spawns an TPU/XLA process and only uses chip 1, Process B, C, etc can not access any TPU device, whether that's chip 1 or 3 other ones not being used.

This then ties into other problems now:
- `torch_xla.distributed.xla_multiprocessing.spawn` spawns 4 processes per chip, which can access the 2 device of that chip => making it harder to communicate and synchronise accross the devices, and in turn harder for parallelism. I will have to look more into how their SPMD/FSDPv2 works, this will definitely give me a better sense how I might solve these issues.

Other than this, I haven't had any resource/help from the xla team so it's a bit difficult to work with. Is there any other easier and faster channel in which I can contact and request DeepSpeed's help from?

**Comment by tjruwase** [comment]
@blueviggen can you help?

**Comment by tjruwase** [comment]
> [@tjruwase](https://github.com/tjruwase) I'm currently testing this simple training script here from [DeepSpeedExamples/alexnet](https://github.com/microsoft/DeepSpeedExamples/blob/master/training/pipeline_parallelism/train.py),

For this working example, did you apply accelerator abstraction by eliminating this cuda [reference](https://github.com/microsoft/DeepSpeedExamples/blob/a85b5e6c828819441a994127f2c1bf6e71def25d/training/pipeline_parallelism/train.py#L154)?

**Comment by radna0** [comment]
> > [@tjruwase](https://github.com/tjruwase) I'm currently testing this simple training script here from [DeepSpeedExamples/alexnet](https://github.com/microsoft/DeepSpeedExamples/blob/master/training/pipeline_parallelism/train.py),
> 
> [@radna0](https://github.com/radna0), for this working example, did you apply accelerator abstraction by eliminating this cuda [reference](https://github.com/microsoft/DeepSpeedExamples/blob/a85b5e6c828819441a994127f2c1bf6e71def25d/training/pipeline_parallelism/train.py#L154)?

Yes for any script I work with, calling any cuda reference results in a lacking nvidia driver error

**Comment by tjruwase** [comment]
@radna0, do you have any thoughts on the above [proposal](https://github.com/microsoft/DeepSpeed/issues/6901#issuecomment-2619144362) to bypass deepspeed launcher?

**Comment by radna0** [comment]
> @radna0, unfortunately, we don't have experience/access to xla team or tpu devices. This is the best way to contact us.

@tjruwase  I have some tpu v3-8 which I'm not running at the moment, I would be happy to provide you with them


> Have you considered bypassing deepspeed launcher entirely? I ask for the following reasons:
> 
> 1. DeepSpeed launcher is not a requirement to access DeepSpeed optimizations. The main entry point to DeepSpeed runtime is [deepspeed.initialize()](https://github.com/microsoft/DeepSpeed/blob/46c6c9e6ce9ede0eeb8a909958a4fecd2fbd942e/deepspeed/__init__.py#L69).
> 2. For transformers [integration](https://huggingface.co/docs/transformers/deepspeed), deepspeed launcher is just one of the [deployment](https://huggingface.co/docs/transformers/deepspeed?multinode=torchrun#deployment) options with torchrun seeming to be [favored](https://huggingface.co/docs/transformers/deepspeed?multinode=torchrun#multi-node-deployment).
> 3. From your XLA post, I tracked down to this [transformers usage](https://github.com/AI-Hypercomputer/tpu-recipes/blob/45f6f66faf130bceb91b034bb92de0e72d9c2fe7/training/trillium/Llama3-70B-PyTorch/GCE/train.sh#L31-L32) of fsdp. I believe that both transformers and accelerate have simplified interchanging fsdp and deepspeed, so did you try replacing with `--deepspeed ds_config` option?

This is actually interesting, I wasn't really sure if deepspeed launcher was integral to making it work, thanks for letting me know.

1. I'll test out the training script to see if it works the same way with `deepspeed.initialize()`
2. I tried using torchrun before but it wasn't suitable for what I was doing, my training script requires separated tpus in different regions to work, more like how some of the recent researches are working on decentralized training, Eg from Nous?
3. I haven't tried this, but this relates part to reason 2 and the other being that this implementation usually only supports higher end tpu version, leaving out my machine version for example in most cases.

**Comment by sfc-gh-truwase** [comment]
@radna0, are you still working on this?
